src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,390 @@
+from __future__ import annotations
+
+import copy
+import re
+from collections.abc import Hashable, Iterable
+from typing import Annotated, Any, Literal, Union
+
+try:
+    from typing import Self
+except ImportError:
+    from typing_extensions import Self
+
+from pydantic import BaseModel, Field, RootModel, computed_field, model_validator
+
+StepPair = Annotated[
+    list[Annotated[int, Field(ge=0)]], Field(min_length=2, max_length=2)
+]
+
+
+class UprootFileSpec(BaseModel):
```

Comment:
A better name might be ROOTFileSpec, since ParquetFileSpec is for Parquet files and this is for ROOT files